File: betterrandom/src/test/java/io/github/pr0methean/betterrandom/prng/adapter/ReseedingSplittableRandomAdapterTest.java
package io.github.pr0methean.betterrandom.prng.adapter;

import static io.github.pr0methean.betterrandom.seed.DefaultSeedGenerator.DEFAULT_SEED_GENERATOR;
import static io.github.pr0methean.betterrandom.seed.SecureRandomSeedGenerator.DEFAULT_INSTANCE;
import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertNotEquals;

import com.google.common.testing.SerializableTester;
import io.github.pr0methean.betterrandom.FlakyRetryAnalyzer;
import io.github.pr0methean.betterrandom.prng.BaseRandom;
import io.github.pr0methean.betterrandom.prng.RandomTestUtils;
import io.github.pr0methean.betterrandom.prng.RandomTestUtils.EntropyCheckMode;
import io.github.pr0methean.betterrandom.seed.FakeSeedGenerator;
import io.github.pr0methean.betterrandom.seed.SeedException;
import io.github.pr0methean.betterrandom.seed.SeedGenerator;
import io.github.pr0methean.betterrandom.seed.SemiFakeSeedGenerator;
import io.github.pr0methean.betterrandom.seed.SimpleRandomSeeder;
import io.github.pr0methean.betterrandom.util.BinaryUtils;
import java.util.concurrent.ThreadLocalRandom;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

public class ReseedingSplittableRandomAdapterTest
    extends SingleThreadSplittableRandomAdapterTest {

  private SimpleRandomSeeder thread;

  @BeforeMethod public void setUp() {
    thread = new SimpleRandomSeeder(getTestSeedGenerator());
  }

  @AfterMethod public void tearDown() {
    thread.stopIfEmpty();
    thread = null;
  }

  @Override protected EntropyCheckMode getEntropyCheckMode() {
    return EntropyCheckMode.LOWER_BOUND;
  }

  @Override protected ReseedingSplittableRandomAdapter createRng() throws SeedException {
    return new ReseedingSplittableRandomAdapter(getTestSeedGenerator(), thread);
  }

  @Override protected BaseRandom createRng(byte[] seed) throws SeedException {
    ReseedingSplittableRandomAdapter out = createRng();
    out.setSeed(seed);
    return out;
  }

  // FIXME: Why does this need more time than other PRNGs?!
  @Test(timeOut = 120_000) @Override public void testDistribution() throws SeedException {
    super.testDistribution();
  }

  @Override public void testInitialEntropy() {
    // This test needs a separate instance from all other tests, but createRng() doesn't provide one
    ReseedingSplittableRandomAdapter random = createRng();
    assertEquals(random.getEntropyBits(), Long.SIZE, "Wrong initial entropy");
  }

  // FIXME: Why does this need more time than other PRNGs?!
  @Test(timeOut = 120_000) @Override public void testIntegerSummaryStats() throws SeedException {
    super.testIntegerSummaryStats();
  }

  @Override @Test public void testSerializable() throws SeedException {
    try {
      SimpleRandomSeeder thread = new SimpleRandomSeeder(DEFAULT_SEED_GENERATOR);
      final BaseSplittableRandomAdapter adapter =
          new ReseedingSplittableRandomAdapter(DEFAULT_INSTANCE, thread);
      final BaseSplittableRandomAdapter clone = SerializableTester.reserialize(adapter);
      assertEquals(adapter, clone, "Unequal after serialization round-trip");
    } finally {
      thread.shutDown();
    }
  }

  @Override protected Class<? extends BaseRandom> getClassUnderTest() {
    return ReseedingSplittableRandomAdapter.class;
  }

  @Override @Test(enabled = false) public void testRepeatability() {
    // No-op.
  }

  @Override @Test(enabled = false) public void testRepeatabilityNextGaussian() {
    // No-op.
  }

  @Override @Test(retryAnalyzer = FlakyRetryAnalyzer.class)
  public void testReseeding() {
    SeedGenerator generator = new SemiFakeSeedGenerator(ThreadLocalRandom.current(), "testReseeding");
    SimpleRandomSeeder seeder = new SimpleRandomSeeder(generator);
    try {
      ReseedingSplittableRandomAdapter random = new
          ReseedingSplittableRandomAdapter(generator, seeder);
      RandomTestUtils.checkReseeding(generator, random, false);
    } finally {
      seeder.shutDown();
    }
  }

  /**
   * Test for crashes only, since setSeed is a no-op.
   */
  @Override @Test public void testSetSeedAfterNextLong() throws SeedException {
    final BaseRandom prng = createRng();
    prng.nextLong();
    prng.setSeed(getTestSeedGenerator().generateSeed(8));
    prng.setSeed(BinaryUtils.convertBytesToLong(getTestSeedGenerator().generateSeed(8)));
    prng.nextLong();
  }

  /**
   * Test for crashes only, since setSeed is a no-op.
   */
  @Override @Test public void testSetSeedAfterNextInt() throws SeedException {
    final BaseRandom prng = createRng();
    prng.nextInt();
    prng.setSeed(getTestSeedGenerator().generateSeed(8));
    prng.setSeed(BinaryUtils.convertBytesToLong(getTestSeedGenerator().generateSeed(8)));
    prng.nextInt();
  }

  /**
   * Assertion-free since reseeding may cause divergent output.
   */
  @Override @Test(timeOut = 10000) public void testSetSeedLong() {
    createRng().setSeed(0x0123456789ABCDEFL);
  }

  /**
   * setRandomSeeder doesn't work on this class and shouldn't pretend to.
   */
  @Override @Test(expectedExceptions = UnsupportedOperationException.class)
  public void testRandomSeederThreadIntegration() {
    SimpleRandomSeeder thread = new SimpleRandomSeeder(DEFAULT_INSTANCE);
    try {
      createRng().setRandomSeeder(thread);
    } finally {
      thread.stopIfEmpty();
    }
  }

  @Test public void testSetSeedGeneratorNoOp() {
    createRng().setRandomSeeder(thread);
  }

  @Override @Test(enabled = false) public void testSeedTooShort() {
    // No-op.
  }

  @Override @Test(enabled = false) public void testSeedTooLong() {
    // No-op.
  }

  @Override @Test public void testDump() throws SeedException {
    SimpleRandomSeeder thread = new SimpleRandomSeeder(DEFAULT_INSTANCE);
    try {
      ReseedingSplittableRandomAdapter firstInstance
          = new ReseedingSplittableRandomAdapter(getTestSeedGenerator(), thread);
      SimpleRandomSeeder otherThread =
          new SimpleRandomSeeder(new FakeSeedGenerator("Different reseeder"));
      try {
        ReseedingSplittableRandomAdapter secondInstance =
            new ReseedingSplittableRandomAdapter(getTestSeedGenerator(), otherThread);
        assertNotEquals(secondInstance.dump(), firstInstance.dump());
      } finally {
        otherThread.shutDown();
      }
    } finally {
      thread.shutDown();
    }
  }

  /**
   * Assertion-free because thread-local.
   */
  @Override @Test public void testThreadSafety() {
    testThreadSafetyVsCrashesOnly(30, functionsForThreadSafetyTest);
  }
}

File: betterrandom/src/test/java/io/github/pr0methean/betterrandom/prng/adapter/ReseedingSplittableRandomAdapterTest.kt
package io.github.pr0methean.betterrandom.prng.adapter

import io.github.pr0methean.betterrandom.TestUtils.assertGreaterOrEqual
import io.github.pr0methean.betterrandom.seed.DefaultSeedGenerator.DEFAULT_SEED_GENERATOR
import org.testng.Assert.assertEquals
import org.testng.Assert.assertNotEquals

import io.github.pr0methean.betterrandom.CloneViaSerialization
import io.github.pr0methean.betterrandom.prng.BaseRandom
import io.github.pr0methean.betterrandom.prng.RandomTestUtils.EntropyCheckMode
import io.github.pr0methean.betterrandom.seed.FakeSeedGenerator
import io.github.pr0methean.betterrandom.seed.RandomSeederThread
import io.github.pr0methean.betterrandom.seed.SeedException
import io.github.pr0methean.betterrandom.seed.SeedGenerator
import io.github.pr0methean.betterrandom.util.BinaryUtils
import java.util.Arrays
import org.testng.annotations.Test

class ReseedingSplittableRandomAdapterTest : SingleThreadSplittableRandomAdapterTest() {

    protected override val entropyCheckMode: EntropyCheckMode
        get() = EntropyCheckMode.LOWER_BOUND

    protected override val classUnderTest: Class<out BaseRandom>
        get() = ReseedingSplittableRandomAdapter::class.java

    @Throws(SeedException::class)
    override fun createRng(): ReseedingSplittableRandomAdapter {
        return ReseedingSplittableRandomAdapter.defaultInstance
    }

    // FIXME: Why does this need more time than other PRNGs?!
    @Test(timeOut = 120000)
    @Throws(SeedException::class)
    override fun testDistribution() {
        super.testDistribution()
    }

    // FIXME: Why does this need more time than other PRNGs?!
    @Test(timeOut = 120000)
    @Throws(SeedException::class)
    override fun testIntegerSummaryStats() {
        super.testIntegerSummaryStats()
    }

    @Test
    @Throws(SeedException::class)
    override fun testSerializable() {
        val adapter = createRng()
        assertEquals(adapter, CloneViaSerialization.clone<BaseSplittableRandomAdapter>(adapter))
    }

    @Test(enabled = false)
    override fun testRepeatability() {
        // No-op.
    }

    @Test(enabled = false)
    override fun testRepeatabilityNextGaussian() {
        // No-op.
    }

    @Test
    override fun testReseeding() {
        val rng = createRng()
        val oldSeed = rng.seed
        while (rng.entropyBits > java.lang.Long.SIZE) {
            rng.nextLong()
        }
        try {
            var newSeed: ByteArray
            do {
                rng.nextBoolean()
                Thread.sleep(10)
                newSeed = rng.seed
            } while (Arrays.equals(newSeed, oldSeed))
            assertGreaterOrEqual(rng.entropyBits, newSeed.size * 8L - 1)
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }

    }

    /** Test for crashes only, since setSeed is a no-op.  */
    @Test
    @Throws(SeedException::class)
    override fun testSetSeedAfterNextLong() {
        val prng = createRng()
        prng.nextLong()
        prng.seed = testSeedGenerator.generateSeed(8)
        prng.setSeed(BinaryUtils.convertBytesToLong(testSeedGenerator.generateSeed(8)))
        prng.nextLong()
    }

    /** Test for crashes only, since setSeed is a no-op.  */
    @Test
    @Throws(SeedException::class)
    override fun testSetSeedAfterNextInt() {
        val prng = createRng()
        prng.nextInt()
        prng.seed = testSeedGenerator.generateSeed(8)
        prng.setSeed(BinaryUtils.convertBytesToLong(testSeedGenerator.generateSeed(8)))
        prng.nextInt()
    }

    /** Assertion-free since reseeding may cause divergent output.  */
    @Test(timeOut = 10000)
    override fun testSetSeedLong() {
        createRng().setSeed(0x0123456789ABCDEFL)
    }

    /**
     * This class manages its own interaction with a RandomSeederThread, so setSeederThread makes no
     * sense.
     */
    @Test(expectedExceptions = arrayOf(UnsupportedOperationException::class))
    @Throws(Exception::class)
    override fun testRandomSeederThreadIntegration() {
        createRng().seedGenerator = testSeedGenerator
    }

    @Test(enabled = false)
    override fun testSeedTooShort() {
        // No-op.
    }

    @Test(enabled = false)
    override fun testSeedTooLong() {
        // No-op.
    }

    @Test
    @Throws(SeedException::class)
    override fun testDump() {
        assertNotEquals(ReseedingSplittableRandomAdapter.getInstance(DEFAULT_SEED_GENERATOR).dump(),
                ReseedingSplittableRandomAdapter.getInstance(testSeedGenerator).dump())
    }

    @Test
    @Throws(SeedException::class)
    fun testFinalize() {
        val generator = FakeSeedGenerator()
        ReseedingSplittableRandomAdapter.getInstance(generator)
        try {
            Runtime.getRuntime().runFinalization()
        } finally {
            System.gc()
            RandomSeederThread.stopIfEmpty(generator)
        }
    }

    /** Assertion-free because thread-local.  */
    @Test
    override fun testThreadSafety() {
        testThreadSafetyVsCrashesOnly(30, BaseRandomTest.FUNCTIONS_FOR_THREAD_SAFETY_TEST)
    }
}
